Use e2epod.WaitForPodNameRunningInNamespace directly

oomichi opened the pull request:
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

WaitForPod*() are just wrapper functions for e2epod package, and they made an invalid dependency to sub e2e framework from the core framework.
So this replaces WaitForPodRunning() with the e2epod function.

Ref: https://github.com/kubernetes/kubernetes/issues/81245

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

oomichi commented:
Circular dependency is made, so we need to move all these functions in a single PR for avoid that.

/hold

oomichi commented on test/e2e/framework/framework.go:
```diff
@@ -55,7 +55,6 @@ import (
 	// TODO: Remove the following imports (ref: https://github.com/kubernetes/kubernetes/issues/81245)
```
This TODO is what this PR does.

neolit123 reviewed (commented):
/lgtm

oomichi commented:
Oops, the other core framework files depend on e2epod package, then circular dependency happens now.
So it is better to make temporary sub e2e package and move these functions to the package for avoiding the dependency temporary.

/retitle WIP: Move WaitForPod*() from e2e core framework

tanjunchen commented on test/e2e/storage/empty_dir_wrapper.go:
```diff
@@ -191,9 +191,9 @@ var _ = utils.SIGDescribe("EmptyDir wrapper volumes", func() {
+	// Slow by design [‾150 Seconds].
 	// This test uses deprecated GitRepo VolumeSource so it MUST not be promoted to Conformance.
-	// To provision a container with a git repo, mount an EmptyDir into an InitContainer that clones the repo using git, then mount the EmptyDir into the Pod’s container.
+	// To provision a container with a git repo, mount an EmptyDir into an InitContainer that clones the repo using git, then mount the EmptyDir into the Pod窶冱 container.
```
EmptyDir into the Pod窶冱 container  ??

oomichi commented:
This is still WIP..

oomichi commented:
/retest

oomichi commented:
/retest

oomichi commented:
/retest

oomichi commented:
/test pull-kubernetes-integration

fejta reviewed (commented):
/lgtm

oomichi commented:
/test pull-kubernetes-integration

oomichi commented:
/hold cancel

alejandrox1 commented on test/e2e/framework/podwaits/podwaits.go:
```diff
@@ -0,0 +1,27 @@
+
+// WaitForPodRunning waits for the pod to run in the namespace.
+func WaitForPodRunning(f *framework.Framework, podName string) error {
+	return e2epod.WaitForPodNameRunningInNamespace(f.ClientSet, podName, f.Namespace.Name)
```
We should move this into the e2epod package and refactor it to use a clienset and a namespace instead of extracting it itself from the framework - to allow people to use it as they wish and to avoid the circular dependency in e2epod.
An example of a function like this one https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/pod/wait.go#L105

```suggestion
func WaitForPodRunning(c clientset.Interface, ns string podName string) error {
	return WaitForPodNameRunningInNamespace(c, podName, ns)
```
The above (i think it should work) can now be included with e2epod.
All calling functions just need to be updated to use it
```go
e2epod.WaitForPodRunning(f.ClientSet, f.Namespace.Name, PodName)
```

oomichi replied:
Yeah, that is a nice point.
I did the same thing by implementing it in e2epod, but that made circular dependency.
Then I think it is nice to move this function into separated package e2epodwaits.
Ideally e2epod is the best, but e2epodwaits is not so bad at this time I feel.

alejandrox1 replied:
What caused the circular dependency in e2epod ?
The only place within the framework where I see this being used in in  test/e2e/framework/framework.go.
If anything in the framework is using that function then we should move it to a subpkg as well.

If we can we should avoid making temp packages like e2epodwaits.
It can make code reviewing easier for us but there is the possibility that people are writing/maintaining their e2e tests and having code be shifted may be confusing.
If we can avoid a temporary subpkg we should, unless there it requires a lot of refactoring work to do so.

oomichi replied:
> What caused the circular dependency in e2epod ?

Because `e2e/framework/util.go` uses e2epod as https://github.com/kubernetes/kubernetes/blob/b6f11382f5d63313b968028719e54a1afb3fe1ba/test/e2e/framework/util.go#L81
That is an invalid dependency(core framework import sub frameworks) as you know, and it is hard to solve it at this time.
In addition, the moved function depends on core framework code as this PR.
So if moving the function into e2epod, circular dependency happens like
- core framework `util.go` imports e2epod (wrongly)
- e2epod becomes importing core framework

> If we can we should avoid making temp packages like e2epodwaits. It can make code reviewing easier for us but there is the possibility that people are writing/maintaining their e2e tests and having code be shifted may be confusing. If we can avoid a temporary subpkg we should, unless there it requires a lot of refactoring work to do so.

Yes, I totally agree with you.
TBH, I am not sure this e2epodwaits is temporary workaround or not because there are already similar subpackages like e2epod and e2epodlogs.

alejandrox1 replied:
> In addition, the moved function depends on core framework code as this PR.

To solve that, `WaitForPodRunning` can be rewritten to not import the framework and instead take a clientset and a namespace as arguments
```go
func WaitForPodRunning(c clientset.Interface, ns string podName string) error {
```

> Because e2e/framework/util.go uses e2epod as

Looking at this PR, the e2e framework doesn't make use of `WaitForPodRunning` so it would not cause a circular dependency. If it was rewritten (I think).
`WaitForPodRunning` depends on code that is well contained within e2epod
* https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/pod/wait.go#L320

---

With respect to fixing the fact that the framework imports e2epods, we need to refactor some more code from utils to not have the framework importing it and that is an AI.
Whenever that gets done it would also be an AI to merge e2epodwaits with e2epod.
I think that if we can do it now there is no reason to leave it for later.
Then we can just focus on removing the e2epod import from framework/util.go

also, just looked at e2epodlogs and I think that one would also be easy to consolidate within e2epod (as an aside thought).

oomichi replied:
@alejandrox1 

> To solve that, WaitForPodRunning can be rewritten to not import the framework and instead take a clientset and a namespace as arguments

That seems a nice idea, thanks! :-)
OK, let me try it.

oomichi replied:
We just need to replace these calls with the existing `e2epod.WaitForPodNameRunningInNamespace(f.ClientSet, podName, f.Namespace.Name)` of e2epod package without moving the function because of specifying the namespace.

oomichi replied:
@alejandrox1 
> also, just looked at e2epodlogs and I think that one would also be easy to consolidate within e2epod (as an aside thought).

I did the above with https://github.com/kubernetes/kubernetes/pull/89180

alejandrox1 reviewed (changes requested):
Thankyou for digging this one up @oomichi !

/priority backlog
/lgtm cancel
/hold

Commit:
Use e2epod.WaitForPodNameRunningInNamespace directly

WaitForPod*() are just wrapper functions for e2epod package, and they
made an invalid dependency to sub e2e framework from the core framework.
So this replaces WaitForPodRunning() with the e2epod function.

oomichi commented:
/test pull-kubernetes-integration

tanjunchen commented:
/retest

oomichi commented:
/test pull-kubernetes-integration

oomichi commented:
@alejandrox1 Could you take a look at this again?
I think I've applied your point on the latest PR.

alejandrox1 reviewed (commented):
/lgtm
/approve

let's get it in before CI complains again :sweat_smile:

alejandrox1 commented:
/hold cancel

oomichi commented:
Thanks for taking a look at this again @alejandrox1 :-)

/test pull-kubernetes-e2e-kind

oomichi commented:
/retest

oomichi commented:
/test pull-kubernetes-integration

oomichi commented:
/test pull-kubernetes-e2e-gce
